File: problems/[E]160.Intersection-of-Two-Linked-Lists.kt
/**
 * Runtime: 184 ms, faster than 35.07% of Kotlin online submissions for Intersection of Two Linked Lists.
 * Memory Usage: 39.1 MB, less than 30.60% of Kotlin online submissions for Intersection of Two Linked Lists.
 */
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun getIntersectionNode(headA:ListNode?, headB:ListNode?):ListNode? {
        val nodesSet = mutableSetOf<ListNode>()
        var nodeA = headA
        var nodeB = headB
        while (nodeA != null || nodeB != null) {
            if (nodeA != null) {
                if (nodeA in nodesSet) {
                    return nodeA
                } else {
                    nodesSet.add(nodeA)
                    nodeA = nodeA.next
                }
            }
            if (nodeB != null) {
                if (nodeB in nodesSet) {
                    return nodeB
                } else {
                    nodesSet.add(nodeB)
                    nodeB = nodeB.next
                }
            }
        }
        
        return null
    }
}

File: swift/[E]160.Intersection-of-Two-Linked-Lists.swift
/**
 * Runtime: 280 ms, faster than 61.96% of Swift online submissions for Intersection of Two Linked Lists.
 * Memory Usage: 16.8 MB, less than 94.02% of Swift online submissions for Intersection of Two Linked Lists.
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func getIntersectionNode(_ headA: ListNode?, _ headB: ListNode?) -> ListNode? {
        var firstPointer = headA
        var secondPointer = headB
        var firstChanged = false
        var secondChanged = false
        while firstPointer != nil && secondPointer != nil && firstPointer !== secondPointer {
            if firstPointer!.next == nil && !firstChanged {
                firstChanged = true
                firstPointer = headB
            } else {
                firstPointer = firstPointer!.next
            }
            if secondPointer!.next == nil && !secondChanged {
                secondChanged = true
                secondPointer = headA
            } else {
                secondPointer = secondPointer!.next
            }
        }
        if firstPointer === secondPointer {
            return firstPointer
        } else {
            return nil
        }
    }
}
